File: android/app/src/main/kotlin/com/goalmaker/app/application/planning/StatsRules.kt
package com.goalmaker.app.application.planning

import java.time.LocalDate

/**
 * The numbers behind the stats screen (docs/stats.md, contracts/vectors/stats.json): tasks finished a
 * week at a time, goals hit a month at a time, how each habit is holding up, and past ratings.
 */
object StatsRules {
    /** How many weeks, months and reviews the screen shows by default. */
    const val WEEKS = 12
    const val MONTHS = 6
    const val RATINGS = 12

    fun build(
        tasks: List<TaskItem>,
        goals: List<GoalItem>,
        entries: List<GoalEntryItem>,
        habits: HabitData,
        reviews: List<ReviewItem>,
        today: LocalDate,
        weekCount: Int = WEEKS,
        monthCount: Int = MONTHS,
        ratingKind: String = ReviewRules.WEEKLY,
        ratingCount: Int = RATINGS,
    ): StatsDigest = StatsDigest(
        weeks = weeks(tasks, today, weekCount),
        months = months(goals, tasks, entries, habits, today, monthCount),
        habits = habits(habits, today, weekCount),
        ratings = ratings(reviews, ratingKind, ratingCount),
    )

    /** Tasks finished in each of the last [count] weeks, oldest first, the last one holding [today]. */
    fun weeks(tasks: List<TaskItem>, today: LocalDate, count: Int = WEEKS): List<StatsDigest.Week> {
        val last = GoalRules.periodStart(GoalHorizon.WEEK, today)
        val first = last.minusWeeks((count - 1).coerceAtLeast(0).toLong())
        val byWeek = tasks.filterNot(TaskItem::deleted)
            .mapNotNull(TaskItem::completedDay)
            .filter { !it.isBefore(first) && !it.isAfter(today) }
            .groupingBy { GoalRules.periodStart(GoalHorizon.WEEK, it) }
            .eachCount()
        return (0 until count.coerceAtLeast(0)).map { step ->
            val start = first.plusWeeks(step.toLong())
            StatsDigest.Week(start, byWeek[start] ?: 0)
        }
    }

    /** The goals of each of the last [count] months, oldest first, and how many of them were hit. */
    fun months(
        goals: List<GoalItem>,
        tasks: List<TaskItem>,
        entries: List<GoalEntryItem>,
        habits: HabitData,
        today: LocalDate,
        count: Int = MONTHS,
    ): List<StatsDigest.Month> {
        val last = today.withDayOfMonth(1)
        val first = last.minusMonths((count - 1).coerceAtLeast(0).toLong())
        val live = goals.filter { !it.deleted && it.status != GoalRules.DROPPED && it.horizon == GoalHorizon.MONTH }
        val tasksByGoal = tasks.filter { !it.deleted && it.goalId != null }.groupBy { it.goalId!! }
        val entriesByGoal = entries.groupBy(GoalEntryItem::goalId)
        return (0 until count.coerceAtLeast(0)).map { step ->
            val start = first.plusMonths(step.toLong())
            val month = live.filter { GoalRules.periodStart(GoalHorizon.MONTH, it.periodStart) == start }
            val hit = month.count { goal ->
                GoalRules.progressOf(goal, tasksByGoal[goal.id].orEmpty(), entriesByGoal[goal.id].orEmpty(), habits).hit
            }
            StatsDigest.Month(start, hit, month.size)
        }
    }

    /** How each habit that is not archived did over the last [weeks] weeks, in the order they are kept. */
    fun habits(habits: HabitData, today: LocalDate, weeks: Int = WEEKS): List<StatsDigest.Habit> {
        val from = GoalRules.periodStart(GoalHorizon.WEEK, today).minusWeeks((weeks - 1).coerceAtLeast(0).toLong())
        return habits.habits.filter { !it.deleted && !it.archived }.map { habit ->
            val checkins = habits.checkinsOf(habit.id)
            val pauses = habits.pausesOf(habit.id)
            val states = HabitRules.periodsBetween(habit, from, today).map { HabitRules.state(habit, it, today, checkins, pauses) }
            StatsDigest.Habit(
                id = habit.id,
                name = habit.name,
                emoji = habit.emoji,
                met = states.count { it == HabitPeriodState.MET },
                periods = states.count { it != HabitPeriodState.NONE },
                streak = HabitRules.streak(habit, today, checkins, pauses),
                best = best(states),
            )
        }
    }

    /** The ratings of the last [count] reviews of [kind] that rated anything, oldest first. */
    fun ratings(reviews: List<ReviewItem>, kind: String = ReviewRules.WEEKLY, count: Int = RATINGS): List<StatsDigest.Rating> =
        reviews.filter { !it.deleted && it.kind == kind && (it.mood != null || it.energy != null) }
            .sortedBy(ReviewItem::periodStart)
            .takeLast(count.coerceAtLeast(0))
            .map { StatsDigest.Rating(it.periodStart, it.mood, it.energy) }

    // The longest run of met periods: a missed one ends a run, the rest are passed over like a streak.
    private fun best(states: List<HabitPeriodState>): Int {
        var best = 0
        var run = 0
        states.forEach { state ->
            when (state) {
                HabitPeriodState.MET -> {
                    run++
                    if (run > best) best = run
                }
                HabitPeriodState.MISSED -> run = 0
                else -> Unit
            }
        }
        return best
    }
}

File: windows/src/GoalMaker.Core/Planning/StatsRules.cs
namespace GoalMaker.Core.Planning;

/// <summary>
/// The numbers behind the stats screen (docs/stats.md, contracts/vectors/stats.json): tasks finished a
/// week at a time, goals hit a month at a time, how each habit is holding up, and past ratings.
/// </summary>
public static class StatsRules
{
    /// <summary>How many weeks the tasks and habits cover by default.</summary>
    public const int Weeks = 12;

    /// <summary>How many months the goals cover by default.</summary>
    public const int Months = 6;

    /// <summary>How many past reviews the mood and energy chart holds.</summary>
    public const int Ratings = 12;

    public static StatsDigest Build(
        IReadOnlyList<TaskItem> tasks,
        IReadOnlyList<GoalItem> goals,
        IReadOnlyList<GoalEntryItem> entries,
        IReadOnlyList<HabitItem> habits,
        IReadOnlyList<HabitCheckin> checkins,
        IReadOnlyList<HabitPause> pauses,
        IReadOnlyList<ReviewItem> reviews,
        DateOnly today,
        int weekCount = Weeks,
        int monthCount = Months,
        string ratingKind = ReviewRules.Weekly,
        int ratingCount = Ratings) => new()
        {
            Weeks = WeeksDone(tasks, today, weekCount),
            Months = MonthsHit(goals, tasks, entries, habits, checkins, today, monthCount),
            Habits = HabitRows(habits, checkins, pauses, today, weekCount),
            Ratings = RatingRows(reviews, ratingKind, ratingCount),
        };

    /// <summary>Tasks finished in each of the last <paramref name="count"/> weeks, oldest first, the last one holding today.</summary>
    public static IReadOnlyList<StatsDigest.Week> WeeksDone(IReadOnlyList<TaskItem> tasks, DateOnly today, int count = Weeks)
    {
        var last = GoalRules.PeriodStart(GoalHorizon.Week, today);
        var wanted = Math.Max(count, 0);
        var first = last.AddDays(-7 * Math.Max(wanted - 1, 0));
        var byWeek = tasks
            .Where(task => !task.Deleted)
            .Select(task => task.CompletedDay)
            .Where(day => day is { } finished && finished >= first && finished <= today)
            .GroupBy(day => GoalRules.PeriodStart(GoalHorizon.Week, day!.Value))
            .ToDictionary(group => group.Key, group => group.Count());
        return [.. Enumerable.Range(0, wanted).Select(step =>
        {
            var start = first.AddDays(7 * step);
            return new StatsDigest.Week(start, byWeek.TryGetValue(start, out var done) ? done : 0);
        })];
    }

    /// <summary>The goals of each of the last <paramref name="count"/> months, oldest first, and how many were hit.</summary>
    public static IReadOnlyList<StatsDigest.Month> MonthsHit(
        IReadOnlyList<GoalItem> goals,
        IReadOnlyList<TaskItem> tasks,
        IReadOnlyList<GoalEntryItem> entries,
        IReadOnlyList<HabitItem> habits,
        IReadOnlyList<HabitCheckin> checkins,
        DateOnly today,
        int count = Months)
    {
        var wanted = Math.Max(count, 0);
        var first = new DateOnly(today.Year, today.Month, 1).AddMonths(-Math.Max(wanted - 1, 0));
        var live = goals.Where(goal => !goal.Deleted && goal.Status != GoalRules.Dropped && goal.Horizon == GoalHorizon.Month).ToList();
        var tasksByGoal = tasks.Where(task => !task.Deleted && task.GoalId is not null).ToLookup(task => task.GoalId!, StringComparer.Ordinal);
        var entriesByGoal = entries.ToLookup(entry => entry.GoalId, StringComparer.Ordinal);
        return [.. Enumerable.Range(0, wanted).Select(step =>
        {
            var start = first.AddMonths(step);
            var month = live.Where(goal => GoalRules.PeriodStart(GoalHorizon.Month, goal.PeriodStart) == start).ToList();
            var hit = month.Count(goal => GoalRules.ProgressOf(goal, tasksByGoal[goal.Id], entriesByGoal[goal.Id], habits, checkins).Hit);
            return new StatsDigest.Month(start, hit, month.Count);
        })];
    }

    /// <summary>How each habit that is not archived did over the last <paramref name="weeks"/> weeks, in the order they are kept.</summary>
    public static IReadOnlyList<StatsDigest.Habit> HabitRows(
        IReadOnlyList<HabitItem> habits,
        IReadOnlyList<HabitCheckin> checkins,
        IReadOnlyList<HabitPause> pauses,
        DateOnly today,
        int weeks = Weeks)
    {
        var from = GoalRules.PeriodStart(GoalHorizon.Week, today).AddDays(-7 * Math.Max(weeks - 1, 0));
        return [.. habits.Where(habit => !habit.Deleted && !habit.Archived).Select(habit =>
        {
            var own = checkins.Where(checkin => checkin.HabitId == habit.Id).ToList();
            var rests = pauses.Where(pause => pause.HabitId == habit.Id).ToList();
            var states = HabitRules.PeriodsBetween(habit, from, today).Select(start => HabitRules.State(habit, start, today, own, rests)).ToList();
            return new StatsDigest.Habit(
                habit.Id,
                habit.Name,
                habit.Emoji,
                states.Count(state => state == HabitPeriodState.Met),
                states.Count(state => state != HabitPeriodState.None),
                HabitRules.Streak(habit, today, own, rests),
                Best(states));
        })];
    }

    /// <summary>The ratings of the last <paramref name="count"/> reviews of <paramref name="kind"/> that rated anything, oldest first.</summary>
    public static IReadOnlyList<StatsDigest.Rating> RatingRows(IReadOnlyList<ReviewItem> reviews, string kind = ReviewRules.Weekly, int count = Ratings)
    {
        var rated = reviews
            .Where(review => !review.Deleted && review.Kind == kind && (review.Mood is not null || review.Energy is not null))
            .OrderBy(review => review.PeriodStart)
            .ToList();
        var wanted = Math.Max(count, 0);
        return [.. rated.Skip(Math.Max(rated.Count - wanted, 0)).Select(review => new StatsDigest.Rating(review.PeriodStart, review.Mood, review.Energy))];
    }

    // The longest run of met periods: a missed one ends a run, the rest are passed over like a streak.
    private static int Best(IEnumerable<HabitPeriodState> states)
    {
        var best = 0;
        var run = 0;
        foreach (var state in states)
        {
            if (state == HabitPeriodState.Met)
            {
                run++;
                best = Math.Max(best, run);
            }
            else if (state == HabitPeriodState.Missed)
            {
                run = 0;
            }
        }

        return best;
    }
}
